clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -1258,6 +1258,61 @@ def ComplexImagPtrOp : CIR_Op<"complex.imag_ptr", [Pure]> {
   let hasVerifier = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// ComplexBinOp
+//===----------------------------------------------------------------------===//
+
+def ComplexBinOpKind : I32EnumAttr<
+    "ComplexBinOpKind",
+    "complex number binary operation kind",
+    [BinOpKind_Add, BinOpKind_Sub, BinOpKind_Mul, BinOpKind_Div]> {
+  let cppNamespace = "::mlir::cir";
+}
+
+def ComplexRangeKind_Full : I32EnumAttrCase<"Full", 1, "full">;
+def ComplexRangeKind_Improved : I32EnumAttrCase<"Improved", 2, "improved">;
+def ComplexRangeKind_Promoted : I32EnumAttrCase<"Promoted", 3, "promoted">;
+def ComplexRangeKind_Basic : I32EnumAttrCase<"Basic", 4, "basic">;
+def ComplexRangeKind_None : I32EnumAttrCase<"None", 5, "none">;
+
+def ComplexRangeKind : I32EnumAttr<
+    "ComplexRangeKind",
+    "complex multiplication and division implementation",
+    [ComplexRangeKind_Full, ComplexRangeKind_Improved,
+     ComplexRangeKind_Promoted, ComplexRangeKind_Basic,
+     ComplexRangeKind_None]> {
+  let cppNamespace = "::mlir::cir";
+}
+
+def ComplexBinOp : CIR_Op<"complex.binop",
+    [Pure, SameTypeOperands, SameOperandsAndResultType]> {
```

Comment:
I'm a bit on the fence here, my understanding is that the range stuff can only be used with mul and div, right? If so, perhaps add and sub should just be regular binops? We'd need to (1) change the existing binop verifier to choke on complex types out of sub and add (2) Make ComplexRangeKind mandatory in this new op, and you can also remove the complex.binop verifier for it.

There are also no tests that check for the range stuff output, can you please add that for each?